rust-mode.el

```diff
@@ -71,17 +71,19 @@ instead of `prog-mode'. This option requires emacs29+."
     map)
   "Keymap for Rust major mode.")
 
-(if (and (version<= "29.1" emacs-version) rust-mode-treesitter-derive)
-    (require 'rust-mode-treesitter)
-  (require 'rust-prog-mode))
-
 ;;;###autoload
 (autoload 'rust-mode "rust-mode" "Major mode for Rust code." t)
 
 ;;;###autoload
 (add-to-list 'auto-mode-alist '("\\.rs\\'" . rust-mode))
 
 (provide 'rust-mode)
+
+(if (and rust-mode-treesitter-derive
+         (version<= "29.1" emacs-version))
```

Comment:
Emacs > 29.1 can be compiled without tree sitter feature.
We should maybe replace this with `(functionp 'treesit-parser-p)`, or something similar ?

Reply:
That's why this is opt in feature, by default `rust-mode-treesiter-derive` is `nil`. 

> We should maybe replace this with (functionp 'treesit-parser-p), or something similar ?

Even then it should be opt-in IMO (atleast for a couple of years) and that already introduces a new defcustom variable. Also, in my experience I have found `treesit-parser-p` not reliable where during startup it returned nil.

Reply:
I mean we should replace `(version<= "29.1" emacs-version)` with `(functionp 'treesit-parser-p)`.

>  Also, in my experience I have found treesit-parser-p not reliable where during startup it returned nil.

`(functionp 'treesit-parser-p)` doesn't call `treesit-parser-p`, it will just return nil when emacs has not been compiled with treesit (`--without-tree-sitter` flag at compile time)

Reply:
> I mean we should replace (version<= "29.1" emacs-version) with (functionp 'treesit-parser-p).

Ah, I see what you mean. Probably you can open a separate PR for it ?

Reply:
And also the `if` check in rust-mode-treesitter.el

Reply:
Yeah, agree. I have opened an issue to track this: https://github.com/rust-lang/rust-mode/issues/537